contract/test/dns/DNS.ts

```diff
@@ -0,0 +1,354 @@
+      expect(await this.publicResolver.dnsRecord(TestNode, TestSubdomainADnsHash, Constants.DNSRecordType.A)).to.equal('0x' + TestSubdomainAInitialRecord)
+      expect(await this.publicResolver.dnsRecord(TestNode, TestSubdomainBDnsHash, Constants.DNSRecordType.A)).to.equal('0x' + TestSubdomainBInitialRecord1 + TestSubdomainBInitialRecord2)
+      expect(await this.publicResolver.dnsRecord(TestNode, oneNameHash, Constants.DNSRecordType.CNAME)).to.equal('0x' + TestSubdomainOneInitialCnameRecord)
+      expect(await this.publicResolver.dnsRecord(TestNode, TestDomainFqdnHash, Constants.DNSRecordType.SOA)).to.equal('0x' + InitialSoaRecord)
+    })
+
+    it('DNS-002 should update existing records', async function (this: Context) {
+      // a.test.country. 3600 IN A 4.5.6.7 (changing address)
+      const [updatedARecord] = dns.encodeARecord(TestSubdomainA, '4.5.6.7')
+      // country. 86400 IN SOA ns1.countrydns.xyz. hostmaster.test.country. 2018061502 15620 1800 1814400 14400 (changing serial)
+      const [updatedSoaRecord] = dns.encodeSOARecord(TestDomainFqdn, { ...DefaultSoa, serial: 2018061502 })
+      const updatedRecord = '0x' + updatedARecord + updatedSoaRecord
+      await this.publicResolver.connect(this.alice).setDNSRecords(TestNode, updatedRecord)
+      expect(await this.publicResolver.dnsRecord(TestNode, TestSubdomainADnsHash, Constants.DNSRecordType.A)).to.equal('0x' + updatedARecord)
+      expect(await this.publicResolver.dnsRecord(TestNode, TestDomainFqdnHash, Constants.DNSRecordType.SOA)).to.equal('0x' + updatedSoaRecord)
+
+      // unchanged records still exist
+      // b.test.country. 3600 IN A 2.3.4.5
+      const [initB1Rec] = dns.encodeARecord(TestSubdomainB, '2.3.4.5')
+      // b.test.country. 3600 IN A 3.4.5.6
+      const [initB2Rec] = dns.encodeARecord(TestSubdomainB, '3.4.5.6')
+      expect(await this.publicResolver.dnsRecord(TestNode, TestSubdomainBDnsHash, Constants.DNSRecordType.A)).to.equal('0x' + initB1Rec + initB2Rec)
+    })
+
+    it('DNS-003 should keep track of entries', async function (this: Context) {
+      // c.test.country. 3600 IN A 1.2.3.4
+      const [cRec] = dns.encodeARecord(TestSubdomainC, '1.2.3.4')
+      const rec = '0x' + cRec
+      await this.publicResolver.connect(this.alice).setDNSRecords(TestNode, rec)
+
+      // Initial check
+      let hasEntries = await this.publicResolver.hasDNSRecords(TestNode, TestSubdomainCDnsHash)
+      expect(hasEntries).to.be.true
+      hasEntries = await this.publicResolver.hasDNSRecords(TestNode, TestSubdomainDDnsHash)
+      expect(hasEntries).to.be.false
+
+      // Update with no new data makes no difference
+      await this.publicResolver.connect(this.alice).setDNSRecords(TestNode, rec)
+      hasEntries = await this.publicResolver.hasDNSRecords(TestNode, TestSubdomainCDnsHash)
+      expect(hasEntries).to.be.true
+
+      // c.test.country. 3600 IN A
+      const [cRec2] = dns.encodeARecord(TestSubdomainC, '')
```

Comment:
This won't clear the record, so the test below would fail. The reason is that the library will still write 4 bytes of 0-short-integer for empty string input. See dns-js dnsrecord.js 

```
function writeA(out, ip) {
  var parts = ip.split('.');
  for (var i = 0; i < 4; i++) {
    out.byte(parts[i]);
  }
}
```

and 

```
BufferWriter.prototype.byte = function (v) {
  this.buf.writeUInt8(v, this.offset);
  this.offset += 1;
  return this;
};
```